### ISSUE TITLE
SC 332 -- Form input label and helper text

### ISSUE THREAD

**Issue by damiansian** [newjersey/accessibility-pm#19]
## URL
https://demo.fileyourstatetaxes.org/en/questions/phone-number

## Issue
The current label of the text input is not creating a consistent accessible name. There are two `p` tags in the `label` tag. This is causing errors on VoiceOver likely due to the fact we're using `block level` elements inside an `inline` element. 

### P tags in the label tag
```html
<label for="state_file_phone_number_form_phone_number">
  <p class="form-question">Your phone number</p>
<p class="text--help">Enter numbers only</p>
</label>
``` 

## Screen shot
![Image](https://github.com/user-attachments/assets/8ac30086-df7b-4459-b664-f91e7c01603f)

## Video walkthrough
https://github.com/user-attachments/assets/bf48aa16-ae5e-4ff4-9a60-8c50d094b2b7

> [!Note]
> The `aria-required="true"` was added after creating this screen recording. 

## WCAG Success Criterion
[SC 3.3.2: Labels or Instructions (Level A)](https://w3.org/TR/WCAG22#labels-or-instructions)

## Impacted users
Screen reader users rely on labels and instructions to understand how to fill out form elements properly. 

## Suggested fix
This suggestion fixes the issues on Mac and PC screen readers and brings the form element up to full WCAG conformance. There are other issues solved for here that have separate bugs filed. 

What changed: 
- Added autocomplete (SC 1.3.5)
- Associated helper text (SC 3.3.2)
- Added required since skipping entry causes and error (SC 3.3.1 and SC 3.3.2)
- Used `aria-describedby` for helper text AND error text. 

### Current
```html
<label for="state_file_phone_number_form_phone_number"><p class="form-question">Your phone number</p>
<p class="text--help">Enter numbers only</p>
</label>
<input 
    autocomplete="off" 
    autocorrect="off" 
    autocapitalize="off" 
    spellcheck="false" 
    type="text" 
    class="form-width--long text-input" 
    id="state_file_phone_number_form_phone_number" 
    name="state_file_phone_number_form[phone_number]">

    </div>
``` 

  
### Should be:
  ```html
<label id="label" for="phone">Your phone number</label>
<input 
    type="text" 
    name="phone" 
    id="phone" 
    autocomplete="tel" 
    aria-describedby="helper"
    aria-required="true">
<p id="helper">Enter numbers only</p>
</div>
```

**Comment by damiansian** [comment]
Yes, that will work. I just tested and the screen reader picks up the correct semantic role. 

```html
<button role="link">Continue</button>
```

**Comment by mmazanec22** [comment]
Note to self: see also box 14 fields

**Comment by mmazanec22** [comment]
@damiansian this one is in code review. If you've got time to review it, here's the link: https://statefile.pr-5474.getyourrefund-testing.org/en/nj/landing-page

How to test:
- Check out the contact info screens you flagged
- Choose the O Neal Walker catchall persona
- Click on the w2 editing button when you get to that screen; there were some help text changes
- Put a value above $200 in one of the box 14 fields and try to save (this should cause an error)

**Comment by damiansian** [comment]
Contact screen flagged has been fixed: 

<img width="1377" alt="Using aria describedby." src="https://github.com/user-attachments/assets/8f794654-5a27-46cf-afb1-0d41606b814a" />

**Comment by damiansian** [comment]
W2 editing example is good to go: 

<img width="761" alt="VoiceOver reading helper text via aria describedby." src="https://github.com/user-attachments/assets/694d1fc4-6e91-4588-adba-4318fb3e8275" />

**Comment by damiansian** [comment]
Error text is read via `aria-describedby`. 

<img width="726" alt="error text read via aria describedby." src="https://github.com/user-attachments/assets/09fff45b-f092-40a3-bbed-59780190a347" />

> [!NOTE]  
> I would love to hear the word "Error" in the error text. I will create another ticket as to not keep this one open. The red triangle is visible but not announced.

**Comment by mmazanec22** [comment]
Thanks! I'm moving it back to "in review" because it's not merged yet, as CfA is worried about the risk of changing all of our inputs at once. I did my best to write about the user impact to explain why we prioritized this ticket and proposed these changes: https://github.com/codeforamerica/vita-min/pull/5474#issuecomment-2625475684.

Double checking with you @damiansian because I don't want to misrepresent the priority level of this ticket: would it be fair to say that without the p/span tag change, the labels are currently not perceivable at all to some assistive tech users? And that's why you marked it as a severity 2?

**Comment by damiansian** [comment]
@mmazanec22 

Here is the rubric I use for assigning severity... 

![Severity rubric detailed in text below](https://github.com/user-attachments/assets/32a49019-c61c-40a3-9939-572e45cb984c)

This is non-conforming to WCAG but not a full blocker (i.e., not Sev 1). There is a workaround available in that I can access the text in "browse mode" reading of the page. It is a core workflow in that I cannot advance to the next screen without inputing text or making a selection. As it's a core workflow, but there is a workaround, I assigned sev 2. 

It's important to note that this is my opinion based on my understanding of your product. My views are not intended to be hegemonic, and are instead open for interpretation and overruling by the product team.

**Comment by mmazanec22** [comment]
@damiansian this one just got merged into staging after some iteration on the implementation. If you have a moment to retest, that would be lovely! https://demo.fileyourstatetaxes.org/en/nj/landing-page

**Comment by tristanvanech** [comment]
@damiansian sorry for the double ping - are you able to retest this one?

**Comment by damiansian** [comment]
@mmazanec22 and @tristanvanech 

The immediate issue of associating the helper text is solved. However, there is no error text associated with the "your email address" input, though, like we have on other inputs. 

<img width="749" alt="No error text" src="https://github.com/user-attachments/assets/e54acf2e-84ca-482d-8d9d-8aba49998888" />


In another area of the experience, we have explicit error warnings using `aria-describedby`. Additionally, the red error icon is not read by a screen reader. You can either start error text with the word "Error" or use sr-only text. 
<img width="499" alt="Image" src="https://github.com/user-attachments/assets/7fe3ebb9-b568-45f1-9408-187d19a19fe8" />

I'm okay to close this ticket as is. If you'd like me to open another ticket for the error handling, let me know.

**Comment by tristanvanech** [comment]
@damiansian yes another ticket would be great!
